src/file/source/file.rs

```diff
@@ -59,9 +59,10 @@ impl FileSourceFile {
                 )))
             };
         }
-        // Adding a dummy extension will make sure we will not override secondary extensions, i.e. "file.local"
-        // This will make the following set_extension function calls to append the extension.
-        let mut filename = add_dummy_extension(filename);
+
+        // Append an extension that will be replaced when calling `set_extension()`:
+        let mut filename = filename;
+        filename.as_mut_os_string().push(".ext_placeholder");
```

Comment:
Alternatively, I think using a conditional communicates the intention better 😅 (_even though it's safe to just append always in our case_)

```suggestion
        let mut filename = filename;
        // Avoid accidental inference of an extension by appending a placeholder for `set_extension()` to replace (ref: #306)
        if filename.extension().is_some() {
          filename.as_mut_os_string().push(".placeholder");
        }
```

Comment with original PR reference of the issue for added context.

---

I'm not wanting to bike shed this though, so feel free to prefer your suggestion or the original and I'm happy to go with either of those too 👍 (_I personally did not find the `file.local` example helpful, it's definitely not about existing extensions so much as it is a file name with one or more `.` getting it's final suffix replaced accidentally_)

Reply:
> I still prefer my version but can go either way on the location.

"either way on location"? Sorry not sure what you meant by that.

Did you mean you would like your version, and are flexible with where it's located in the file? 🤔 (_presumably because my last suggestion moved the comment under `let mut filename` instead of before it as it once was_)

I'll assume that and go with your comment 👍 

---

The remainder of this response is just sharing my opinion on why I think my version was simpler to grok (_with minimal familiarity required to the reader_).

> When you have a file stem with a `.` it can be mistaken for an extension

It's technically not the file stem though, it's the full PathBuf 😅

If we used `filename.file_stem()` we'd get the stem, but with that same concern of the last `.` inferring an extension. So you're right in that at this point we want to think of `filename` as the file stem with the placeholder appended as the extension to be replaced.

The conditional statement added now better communicates this expectation of a preventative measure for an inferred extension? Which the associated comment provides additional clarity.

---

> ```rust
> // Preserve any extension-like text within the provided file stem (e.g. `file.local`) by appending a fake extension which will be replaced by `set_extension()` calls (e.g. `file.local.json`).
> ```

This reads with more mental overhead to me at least 😅
- "file stem" `file.local` implicitly referring to `filename`
- `file.local.json` as example for output after a `set_extension()` call replaces the fake extension.

> ```rust
> // Adding a dummy extension will make sure we will not override secondary extensions, i.e. "file.local"
> ```

This original comment didn't explain the "why" that well and I had to use `git blame` to lookup the commit/PR for extra context.

This one reads like `myfile.tar` as a 2nd extension to `myfile.tar.gz`, but I don't really think of extensions in a filename liberally using `.` either like `my.file.name.here` 😅

Accidental inference of an extension reads much more clearly (_it's only a 2ndry extension in the context of intent to later append the config extensions via `set_extension()`_).

Reply:
This has swapped out my comment for yours, with a slight modification to better convey the expected transformation for a file stem `file.local`. Hopefully a good compromise 😅

```suggestion
        let mut filename = filename;
        // Preserve any extension-like text within the provided file stem by appending a fake extension 
        // which will be replaced by `set_extension()` calls (e.g.  `file.local.placeholder` => `file.local.json`)
        if filename.extension().is_some() {
          filename.as_mut_os_string().push(".placeholder");
        }
```

I'll commit and squash this, but I don't mind ammending the change again if you're not happy with it 👍